### ISSUE TITLE
Bug concerning the updating of $+

### ISSUE THREAD

**Issue by florian-pe** [Perl/perl5#20401]
So I was trying to write a json parser with a regex (using `local` and the embeded code feature), when I found the bug.
I couldn't golf it meaningfully further.
```
my $json_string = '{"key": "value"}';

$json_string =~ m{
    ^ \s* (?&value) \s* $
    (?(DEFINE)
        (?<value>
                ("\w+")
                (?{
                    say "VALUE ", $+;
                })
            |
                \{  \s* ("\w+") \s* : \s* (?&value)  \s* \}
        )
     )
}x;
```
The subpattern "value" defined here match either a double quoted string, or a json object,and recursion occurs at the point where we expect a json value.
At the point where the embeded code fire, we just matched `"value"`, so `$+` should be set to `"value"` but for some reason it isn't the case, and instead, it contains the value of the previous capturing group, ie `"key"`.

Output:
```
VALUE "key"
```

```
$ perl -V
Summary of my perl5 (revision 5 version 36 subversion 0) configuration:
   
  Platform:
    osname=linux
    osvers=5.12.15-arch1-1
    archname=x86_64-linux-thread-multi
    uname='archlinux'
    config_args='-des -Dusethreads -Duseshrplib -Doptimize=-march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -flto -Dprefix=/usr -Dvendorprefix=/usr -Dprivlib=/usr/share/perl5/core_perl -Darchlib=/usr/lib/perl5/5.36/core_perl -Dsitelib=/usr/share/perl5/site_perl -Dsitearch=/usr/lib/perl5/5.36/site_perl -Dvendorlib=/usr/share/perl5/vendor_perl -Dvendorarch=/usr/lib/perl5/5.36/vendor_perl -Dscriptdir=/usr/bin/core_perl -Dsitescript=/usr/bin/site_perl -Dvendorscript=/usr/bin/vendor_perl -Dinc_version_list=none -Dman1ext=1perl -Dman3ext=3perl -Dcccdlflags='-fPIC' -Dlddlflags=-shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -flto -Dldflags=-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -flto'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -flto'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='12.1.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags ='-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -flto -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib
    libs=-lpthread -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=/lib/../lib/libc.so.6
    so=so
    useshrplib=true
    libperl=libperl.so
    gnulibc_version='2.35'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib/perl5/5.36/core_perl/CORE'
    cccdlflags='-fPIC'
    lddlflags='-shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -flto -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
    USE_THREAD_SAFE_LOCALE
  Built under linux
  Compiled at May 29 2022 08:49:34
  @INC:
    /usr/lib/perl5/5.36/site_perl
    /usr/share/perl5/site_perl
    /usr/lib/perl5/5.36/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib/perl5/5.36/core_perl
    /usr/share/perl5/core_perl
```

**Comment by florian-pe** [comment]
Is this bug considered irrelevant?

**Comment by AnFunctionArray** [comment]
@florian-pe Use `$^N` for the last match instead.

Or alternatively you can name your match and use that as well.

**Comment by demerphq** [comment]
On Sat, 15 Oct 2022 at 17:42, florian-pe ***@***.***> wrote:

> So I was trying to write a json parser with a regex (using local and the
> embeded code feature), when I found the bug.
>

Ok. Not exactly recommended since there are proper modules for this. But ok.


> I couldn't golfed meaningfully further.
>
> my $json_string = '{"key": "value"}';
>
> $json_string =~ m{
>     ^ \s* (?&value) \s* $
>     (?(DEFINE)
>         (?<value>
>                 ("\w+")
>                 (?{
>                     say "VALUE ", $+;
>                 })
>             |
>                 \{  \s* ("\w+") \s* : \s* (?&value)  \s* \}
>         )
>      )
> }x;
>
> The subpattern "value" defined here match either a double quoted string,
> or a json object,and recursion occurs at the point where we expect a json
> value.
> At the point where the embeded code fire, we just matched "value", so $+
> should be set to "value" but for some reason it isn't the case, and
> instead, it contains the value of the previous capturing group.
>

It is not clear to me that this is a bug.  The docs say this:

$LAST_PAREN_MATCH
    $+      The text matched by the highest used capture group of the last
            successful search pattern. It is logically equivalent to the
            highest numbered capture variable ($1, $2, ...) which has a
            defined value.

    $LAST_SUBMATCH_RESULT
    $^N     The text matched by the used group most-recently closed (i.e.
            the group with the rightmost closing parenthesis) of the last
            successful search pattern. This is subtly different from $+. For
            example in

                "ab" =~ /^((.)(.))$/

            we have

                $1,$^N   have the value "ab"
                $2       has  the value "a"
                $3,$+    have the value "b"

            This is primarily used inside "(?{...})" blocks for examining
            text recently matched. For example, to effectively capture text
            to a variable (in addition to $1, $2, etc.), replace "(...)"
            with

                (?:(...)(?{ $var = $^N }))

            By setting and then using $var in this way relieves you from
            having to worry about exactly which numbered set of parentheses
            they are.

            This variable was added in Perl v5.8.0.

If you read this carefully and then look at your pattern the highest
capture buffer with a match is $2 which is the "key" part of the pattern.
When it recurses it ends up rematching $1, which si not the highest buffer
defined in the pattern.

Also it appears that $^N is documented to be exactly for the use case you
are trying to achieve.

I will dig a bit longer, but at this time I do not think this is a bug.


Cheers,
Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by demerphq** [comment]
On Sun, 6 Nov 2022 at 09:44, AnFunctionArray ***@***.***> wrote:
>
> @demerphq So this is a bug then?

Actually, now I am convinced it is not a bug in perl. Try this:

$ cat t.pl
my $json_string = '{"key": "value"}';

$json_string =~ m{
    ^ \s* (?&value) \s* $
    (?(DEFINE)
        (?<value>
                ("\w+")
                (?{
                    printf "Plus:'%s' N: '%s' 1: '%s' 2:'%s' 3:'%s'\n",
                     $+//"-", $^N//"-", $1//"-", $2//"-", $3//"-" ;
                })
            |
                \{  \s* ("\w+") \s* : \s* (?&value)  \s* \}
        )
     )
}x;

print "'$&'\n";
__END__
Plus:'"key"' N: '"value"' 1: '-' 2:'"value"' 3:'"key"'
'{"key": "value"}'

Once "key" is matched to $3, then $+ stays set to "key" as $3 is the
highest defined capture buffer.

The value that florian expects is in the recently captured $2, and is
correctly available in $^N.

I will close this as not-a-bug.

Yves

**Comment by demerphq** [comment]
@AnFunctionArray I agree. I am closing this as not-a-bug.

**Comment by demerphq** [comment]
No. We only have a few folks who look at issues like this (me included) and
I only noticed this now.

Yves

**Comment by AnFunctionArray** [comment]
@demerphq So this is a bug then?

Because on https://perldoc.perl.org/perlre

It says:

`$+ returns whatever the last bracket match matched.`

Which means that it returns the regex last lexical (not dynamic match - which is $^N).

Which in this example is correct to be "key".

**Comment by florian-pe** [comment]
@AnFunctionArray @demerphq
Thank you both of you for taking the time. This is indeed not a bug, and I had minsinterpreted the documentation. `$^N` is what is needed in this case. What I didn't get at the time is that `$+` is really a static lookup (logically) of the highest numbered capturing group whereas `$^N` is more dynamic and reads the captured value at the top of the capture stack.

`Ok. Not exactly recommended since there are proper modules for this. But ok.`
I am doing this for sports really, but also because I wanted to make my own parser which would use Tie::IxHash internally, so that serializing a parsed JSON outputs the same input JSON, without changing the order of all the key/value pairs.

I have also found a practical example for matching a json with a regex.
I wrote a little script which take as input an html file saved with a web browser of a youtube search results page, which extract the title and urls of the videos.
The webpage is very heavy (above 4MB). The easiest way to extract this information is to use the embeded json (around 800KB/1MB) in the page, as opposed to obtain the title via the correct `<span></span>` and the url with the correct `href=".."`.

My first try was to parse the webpage with HTML::TreeBuilder::XPath and then use JSON::XS. It worked but was very slow. But then, I used a regex for extracting the json from the html and it was significantly faster.
The regex basically match
`<script\s[^>]+> \s* var \s+ ytInitialData \s* = \s* ((&value))` where the definition of `(?<value>)` is a JSON value.
I've made a little test to compare those.
The first one parses the html with HTML::TreeBuilder::XPath, use xpath (`findnodes()`) to exctract the json, the use JSON::XS to parse the json and finally serialize it and prints it with Data::Dumper. This takes about `5.7 seconds`.
The second test does the same thing, but extract the json with the regex only, and does not parse the html nor use xpath. Tthis test takes about ~~1~~ ~~0.65~~ `0.52 second` (using `++` and `(?> )`).

This is kind of cool I think.
I can post the regex and the test if you are interested.

**Comment by AnFunctionArray** [comment]
@florian-pe That's nice I - can believe that since I've written my own C lexer/compiler using this method - so in contrast to @demerphq I highly encourage using regex for any parsing needs (that's my personal opinion).


The deal is that you don't use `(??{})` which significantly can slow down the regex if it's generating a new regex since it needs recompilation. But `(?{})` is fine.

**Comment by AnFunctionArray** [comment]
@florian-pe I suggest you create a gihub project rather than posting a flying regex here. I may even check it out and post issues I see.

**Comment by demerphq** [comment]
@AnFunctionArray when I commented that this seems an unusual choice I wasn't making a comment on regex parsing at all, I was merely reflecting that something like JSON::XS is implemented in highly efficient C code, and is basically impossible to beat from Perl.  I have spent some time in the past analyzing its performance and its hard to be faster in C than it is at what it does, let alone perl, frankly it is extremely elegantly and efficiently written, something worth reading if you are into that kind of problem.

Having said that generally writing a parser using regex grammars is a bad idea from a performance point of view. It can be very fast when it matches, and then horribly mis-perform when it doesn't, and writing the patterns that fail fast is hard.  I happen to be the person who added most of the named capture and regex recursion support (made possible because it was on top of @iabyn work refactoring the regex engine into a state machine) so I have no bias towards regex "parsing", on the contrary I have done quite a bit to make it easier. :-)

Nevertheless usually a hybrid approach is much better and saner. Use the regex engine to tokenize the input stream, and then write code to "parse" those tokens appropriately. In that module you wouldn't expect to see a rule /in regex/ that expresses something like a dictionary key value tuple. You would expect to see a subroutine that pulls a token from the input, and then based on what that token is applies the appropriate expectations to the following tokens extracted from the input stream. The regex part would  be restricted to patterns that represent the various tokens allowed. JSON isnt very rich, so that means maybe half-dozen token types I'd guess.  Usually parsers of this form use destructive parsing with a `s/$TOKEN_PAT//` pattern anchored to the front of the pattern, and they replace each parsed token with an empty string as they go.  This form of parser avoids traps like the regex engine going quadratic by backtracking to satisfy a complex grammar rule. Sometimes the "tokens" are closer to expression, and involve some grammar like behavior at the regex level, but in general it is the same idea.

Anyway, I certainly am not opposed to using the regex engine for parsing, but I think that often in Perl there is a temptation to have the regex engine do too much, with negative consequences on performance (especially non matching performance) and on maintainability.  Resisting that temptation usually results in a more efficient and easier to maintain parser.
